src/main/java/com/squareup/javapoet/JavaFile.java

```diff
@@ -258,6 +275,83 @@ public Builder addStaticImport(ClassName className, String... names) {
       return this;
     }
 
+    public Builder alwaysQualify(String... simpleNames) {
+      checkArgument(simpleNames != null, "simpleNames == null");
+      for (String name : simpleNames) {
+        checkArgument(
+            name != null,
+            "null entry in simpleNames array: %s",
+            Arrays.toString(simpleNames)
+        );
+        alwaysQualify.add(name);
+      }
+      return this;
+    }
+
+    /**
+     * Call this to always fully qualify any types that would conflict with possibly nested types of
+     * this {@code typeElement}. For example - if the following type was passed in as the
+     * typeElement:
+     *
+     * <pre><code>
+     *   class Foo {
+     *     class NestedTypeA {
+     *
+     *     }
+     *     class NestedTypeB {
+     *
+     *     }
+     *   }
+     * </code></pre>
+     *
+     * <p>
+     * Then this would add {@code "NestedTypeA"} and {@code "NestedTypeB"} as names that should
+     * always be qualified via {@link #alwaysQualify(String...)}. This way they would avoid
+     * possible import conflicts when this JavaFile is written.
+     *
+     * @param typeElement the {@link TypeElement} with nested types to avoid clashes with.
+     * @return this builder instance.
+     */
+    public Builder avoidClashesWithNestedClasses(TypeElement typeElement) {
```

Comment:
Feels like maybe this should be an overload of declaring a superclass or interface. Otherwise when would you use these functions?

Reply:
That could work, how would that be communicated to the JavaFile when it's figuring out imports though?

Reply:
The list would be maintained in each class and then aggregated into a final list when pulling them together into a file.

Reply:
I've updated this to move the tracking of always qualified types to TypeSpecs here: 67d5094. Now the JavaFile just parses them out of all the typespecs in it. As for making them overloads of superclass/superinterface, there's no APIs for those that accept elements currently. I added them to compliment this